services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java

```diff
@@ -114,6 +114,21 @@ public boolean isUseMtlsHokToken() {
         return Boolean.parseBoolean(useUtlsHokToken);
     }
 
+    // KEYCLOAK-9551 Client Credentials Grant generates refresh token
+    // https://tools.ietf.org/html/rfc6749#section-4.4.3
+    public boolean isUseRefreshTokenForClientCredentialsGrant() {
+        // "true" -> we allow use of refresh_token for client_credentials grant by default for backwards compatibility
+        String val = getAttribute(OIDCConfigAttributes.USE_REFRESH_TOKEN_FOR_CLIENT_CREDENTIALS_GRANT, "true");
```

Comment:
Will it be possible that newly created clients will have this "false" by default and won't generate refresh tokens, but the clients migrated from previous version will have this "true" for backwards compatibility?

@stianst @pedroigor Do you agree with the migration behaviour as mentioned above?

@thomasdarimont Regarding your question in the JIRA about version on the representations, the RealmRepresentation class has field "keycloakVersion" . So this can be perhaps used to check if it is migration from older version or not? If the field is not present, we may stick to default behaviour (false). WDYT?

Reply:
I'm good either way because we should aim to remove refresh token completely from the client credentials grant, medium-term.

Reply:
Thanks for the tip about the "keycloakVersion" on the RealmRep... I give this a shot.

Reply:
Would it be possible to change this in another PR? Changing the default together with introducing the functionality might be too much.

Reply:
Sounds good